File: src/main/java/game/Camera/ObstructionDrawer.java
package game.Camera;

import game.Map.Map;
import game.Map.MapChunk;

import java.util.List;

public class ObstructionDrawer {

    private ObstructionVisibility obstructionVisibility;
    private  ObstructionMover obstructionMover;

    public ObstructionDrawer(ObstructionVisibility obstructionVisibility,ObstructionMover obstructionMover){
        this.obstructionVisibility = obstructionVisibility;
        this.obstructionMover = obstructionMover;
    }

    public void drawMap(Map map, Double gameScreenWidth, Double gameScreenHeight, Double offsetAndDeltaY, Double offsetAndDeltaX) {
        drawMapTiles(map.mapChunks,offsetAndDeltaY,offsetAndDeltaX);
        drawItems(map.mapChunks,offsetAndDeltaY,offsetAndDeltaX);
        setMapTileVisibility(map,gameScreenWidth,gameScreenHeight);
        setItemVisibility(map,gameScreenWidth,gameScreenHeight);
    }


    private void setItemVisibility(Map map, Double gameScreenWidth, Double gameScreenHeight) {
        map.mapChunks.forEach(
                mapChunk -> mapChunk.getItems()
                        .forEach(x ->
                                obstructionVisibility.setVisibility(
                                        x.getNode(),
                                        gameScreenWidth,
                                        gameScreenHeight))
        );
    }

    private void setMapTileVisibility(Map map, Double gameScreenWidth, Double gameScreenHeight) {
        map.mapChunks.forEach(
                mapChunk -> mapChunk.getMapTiles()
                        .forEach(x ->
                                obstructionVisibility.setVisibility(
                                        x.getNode(),
                                        gameScreenWidth,
                                        gameScreenHeight))
        );
    }

    private void drawItems(List<MapChunk> mapChunks, Double offsetAndDeltaY, Double offsetAndDeltaX) {
        mapChunks.forEach(
                mapChunk -> mapChunk.getItems()
                        .forEach(item ->
                                obstructionMover.move(
                                        item,
                                        offsetAndDeltaY,
                                        offsetAndDeltaX))
        );
    }

    private void drawMapTiles(List<MapChunk> mapChunks, Double offsetAndDeltaY, Double offsetAndDeltaX) {
        mapChunks.forEach(
                mapChunk -> mapChunk.getMapTiles()
                        .forEach(item ->
                                obstructionMover.move(
                                        item,
                                        offsetAndDeltaY,
                                        offsetAndDeltaX))
        );
    }

}

File: src/main/java/game/Camera/ObstructionDrawer.kt
package game.Camera

import javafx.geometry.Rectangle2D

class ObstructionDrawer(private val obstructionVisibility: ObstructionVisibility, private val obstructionMover: ObstructionMover) {
    fun draw(mapTilesAndItems: List<Rectangle2D>, gameScreenWidth: Double, gameScreenHeight: Double, offsetAndDeltaY: Double, offsetAndDeltaX: Double): List<Rectangle2D> {
        val s = setNodesVisibility(mapTilesAndItems, gameScreenWidth, gameScreenHeight)
        return moveNodes(s, offsetAndDeltaY, offsetAndDeltaX)
    }

    private fun moveNodes(nodes: List<Rectangle2D>, offsetAndDeltaY: Double, offsetAndDeltaX: Double): List<Rectangle2D> {
        return nodes
                .map { obstruction ->
                    obstructionMover.getNodeX(
                            obstruction,
                            offsetAndDeltaX, offsetAndDeltaY)
                }

    }

    private fun setNodesVisibility(nodes: List<Rectangle2D>, gameScreenWidth: Double, gameScreenHeight: Double): List<Rectangle2D> {
        val map = nodes.map { node -> obstructionVisibility.isVisible(node, gameScreenWidth, gameScreenHeight) }

        return map.map { x -> x.get() }
    }
}
